src/mlpack/methods/lsh/objectivefunction.hpp

```diff
@@ -0,0 +1,171 @@
+    //! Data points for y-axis.
+    const arma::mat* yData;
+};
+
+/**
+ * Returns the value of the objective function for some coordinates (alpha,
+ * beta, gamma).
+ * This is the mean squared error for the current parameters or coordinates.
+ */
+double DefaultObjectiveFunction::Evaluate(const arma::mat& coordinates) const
+{
+  // Use extra variables to make code readable.
+  double alpha = coordinates(0, 0);
+  double beta = coordinates(1, 0);
+  double gamma = coordinates(2, 0);
+  double M = (double) NumFunctions();
+
+  // Sum the squared error for each element in yData.
+  double sum = 0;
+  for (size_t i = 0; i < yData->n_elem; ++i)
+  {
+    // Map i to (row, col). Columnwise access of yData.
+    size_t row = i % yData->n_rows;
+    size_t col = (size_t) (i / yData->n_rows); // Integer division (floor).
+
+    // Get the corresponding values.
+    size_t x = (*xData)(row);
+    size_t k = (*kData)(col);
+    double y = (*yData)(row, col);
+
+    // Evaluate (y - a * k ^ b * x ^ c)^2 for the given (x, y) pair.
+    sum += pow(y - alpha * std::pow(k, beta) * std::pow(x, gamma), 2); 
+  }
+
+  // Return the mean of the squared errors.
+  return sum / M;
+}
+
+/**
+ * Stores the gradient of the objective function in gradient. This is the
+ * derivative with respect to (alpha, beta, gamma) evaluated at the current
+ * parameters.
+ */
+void DefaultObjectiveFunction::Gradient(const arma::mat& coordinates,
+                                        arma::mat& gradient) const
+{
+  // Use extra variables to make code readable.
+  double alpha = coordinates(0, 0);
+  double beta = coordinates(1, 0);
+  double gamma = coordinates(2, 0);
+  double M = (double) NumFunctions();
+
+  // Allocate 3x1 matrix for gradient. Set all gradients to 0.
+  gradient.set_size(3, 1);
+  gradient.zeros(3,1);
+
+  // Sum each gradient.
+  for (size_t i = 0; i < yData->n_elem; ++i)
+  {
+    size_t row = i % yData->n_rows;
+    size_t col = (size_t) (i / yData->n_rows); // Integer division.
+    size_t x = (*xData)(row);
+    size_t k = (*kData)(col);
+    double y = (*yData)(row, col);
+
+    // The error for these parameters. Precompute for efficiency.
+    double error = (y - alpha * std::pow(k, beta) * std::pow(x, gamma));
+
+    // The chain rule factor of the product, for each gradient dimension.
+    double alphaChain = 
+      - 2.0 * std::pow(k, beta) * std::pow(x, gamma);
+
+    double betaChain = 
+      - 2.0 * alpha * std::pow(x, gamma) * std::log(k) * std::pow(k, beta);
+
+    double gammaChain =
+      - 2.0 * alpha * std::pow(k, beta) * std::log(x) * std::pow(x, gamma);
+
+    // 3x1 column vector (in matrix form).
+    gradient(0, 0) += error * alphaChain;
+    gradient(1, 0) += error * betaChain;
+    gradient(2, 0) += error * gammaChain;
+  }
+
+  // Return the average of each gradient after the summation is complete.
+  gradient(0, 0) /= ((double) M);
+  gradient(1, 0) /= ((double) M);
+  gradient(2, 0) /= ((double) M);
+}
+
```

Comment:
In `objectivefunction`, I initialize the point randomly. I'm not sure if that also plays a role, since I don't seed the random generator.

Try using
```
$ bin/mlpack_lshmodel -r iris.csv -p 0.5 -v
```
and
```
bin/mlpack_lshmodel -r iris.csv -p 0.6 -v
```
The first should converge to some real values, while the second should "converge" to NaN after 10000 iterations (which I imagine is the default maximum number of iterations).
Both the cost function and the gradients include logarithms, so I suspect we pass negative values somewhere in there. I'll investigate that and let you know.

I think that's also a bug in L_BFGS - shouldn't the iterations stop if objective is NaN?